@Remote bean can't be accessed from a different app, when using Payara5

# Description #
----------

With the latest version of Payara 4 I am able to deploy the following two apps:

1. App1 has the following code(the below interface has single implementation):
```
import javax.ejb.Remote;

@Remote
public interface EmailNotificationManagerRemote {

	public void sendMail(String[] recipientsTo, String[] recipientsCC, String[] recipientsBCC, String subject,
            String body, String senderName, String senderMail, String mimeType, String mailSession, Boolean isSMS)
            throws InvalidInputException;
	
	 public void sendSupportMail(String errorCode, String subject, String body, String logLevel)  throws InvalidInputException;
}
```

App2 has the following code:

```
@EJB(mappedName="EmailNotificationManager")
private EmailNotificationManagerRemote emailNotificationManagerRemote;
```

When deploying App2(after App1 has already been successfully deployed), I get the following exception:

```
Exception during lifecycle processing
org.glassfish.deployment.common.DeploymentException: JNDI lookup failed for the resource: Name: [java:module/env/com.***.KeyLicenseNotifierBean/emailNotificationManagerRemote], Lookup: [EmailNotificationManager#com.***.EmailNotificationManagerRemote], Type: [Session]
...
Caused by: javax.naming.CommunicationException: Communication exception for SerialContext[myEnv={java.naming.factory.state=com.sun.corba.ee.impl.presentation.rmi.JNDIStateFactoryImpl, java.naming.factory.url.pkgs=com.sun.enterprise.naming} [Root exception is java.lang.RuntimeException: Cant copy Serializable object:]
...
Caused by: java.lang.RuntimeException: Cant copy Serializable object:
...
Caused by: java.lang.IllegalArgumentException: Illegal null argument to ObjectInputStreamWithLoader
```

## Expected Outcome

If App1 is successfully deployed, App2 should be deployed without exceptions.

## Current Outcome

App2 deployment fails.

## Steps to reproduce (Only for bug reports) 

Create two wars/jars one with a @Remote EJB definition and one which uses this remote bean. Check that when deploying the app, which uses the bean, the bean can't be found...

## Context (Optional)

The same two wars are deployable and are working on Payara4.

## Environment ##

- **Payara Version**: 5
- **Edition**:  Full
- **JDK Version**:  1.8.0_151
- **Operating System**:  Windows 10.0.15063 Build 15063

I investigate on my side, and for me the pb comes from the fact the contextClassLoader of a DefaultWorkerThread is not correctly initialized or injected (with com.sun.enterprise.util.Utility.setContextClassLoader) after loading ejb remote classe.

stacktrace :

Caused by: java.lang.IllegalArgumentException: Illegal null argument to ObjectInputStreamWithLoader
	at org.glassfish.common.util.ObjectInputStreamWithLoader.<init>(ObjectInputStreamWithLoader.java:69)
	at org.glassfish.common.util.OSGiObjectInputOutputStreamFactoryImpl$OSGiObjectInputStream.<init>(OSGiObjectInputOutputStreamFactoryImpl.java:150)
	at org.glassfish.common.util.OSGiObjectInputOutputStreamFactoryImpl.createObjectInputStream(OSGiObjectInputOutputStreamFactoryImpl.java:136)
	at com.sun.enterprise.naming.util.NamingUtilsImpl.makeCopyOfObject(NamingUtilsImpl.java:130)
	... 47 more


Regards,

--- a/appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java
+++ b/appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java
@@ -37,7 +37,7 @@
  * only if the new code is made subject to such option by the copyright
  * holder.
  */
-// Portions Copyright [2016-2017] [Payara Foundation and/or its affiliates]
+// Portions Copyright [2016-2018] [Payara Foundation and/or its affiliates]
 
 package com.sun.enterprise.deployment.util;
 
@@ -110,7 +110,7 @@ public class ResourceValidator implements EventListener, ResourceValidatorVisito
     private Domain domain;
     
     @Inject
-    JavaEEContextUtil contextUtil;
+    private JavaEEContextUtil contextUtil;
 
     private static LocalStringManagerImpl localStrings = new LocalStringManagerImpl(ResourceValidator.class);
 
@@ -123,7 +123,7 @@ public class ResourceValidator implements EventListener, ResourceValidatorVisito
     }
 
     @Override
-    public void event(Event event) {
+    public void event(Event<?> event) {
         if (event.is(Deployment.AFTER_APPLICATION_CLASSLOADER_CREATION)) {
             DeploymentContext deploymentContext = (DeploymentContext) event.hook();
             Application application = deploymentContext.getModuleMetaData(Application.class);
@@ -139,18 +139,9 @@ public class ResourceValidator implements EventListener, ResourceValidatorVisito
             AppResources appResources = new AppResources();
             // Puts all resources found in the application via annotation or xml into appResources
             parseResources(deploymentContext, application, appResources);
-            
+
             // Ensure we have a valid component invocation before triggering lookups
-            String componentId = null;
-            for (BundleDescriptor bundleDescriptor : application.getBundleDescriptors()) {
-                if (bundleDescriptor instanceof JndiNameEnvironment) {
-                    componentId = DOLUtils.getComponentEnvId((JndiNameEnvironment) bundleDescriptor);
-                    if (componentId != null) {
-                        break;
-                    }
-                }
-            }
-            contextUtil.setInstanceComponentId(componentId);
+            contextUtil.setEmptyInvocation();
             try (Context ctx = contextUtil.pushContext()) {
                 validateResources(deploymentContext, application, appResources);
             }

--- a/appserver/web/gf-web-connector/src/main/java/fish/payara/appserver/context/ContextImpl.java
+++ b/appserver/web/gf-web-connector/src/main/java/fish/payara/appserver/context/ContextImpl.java
@@ -63,9 +63,9 @@ class ContextImpl {
             }
         }
 
-        private final ClassLoader oldClassLoader;
         private final ComponentInvocation invocation;
         private final InvocationManager invMgr;
+        private final ClassLoader oldClassLoader;
     }
 
     @RequiredArgsConstructor

--- a/appserver/web/gf-web-connector/src/main/java/fish/payara/appserver/context/JavaEEContextUtilImpl.java
+++ b/appserver/web/gf-web-connector/src/main/java/fish/payara/appserver/context/JavaEEContextUtilImpl.java
@@ -89,7 +89,6 @@ public class JavaEEContextUtilImpl implements JavaEEContextUtil, Serializable {
      */
     @Override
     public Context pushContext() {
-        ClassLoader oldClassLoader = Utility.getClassLoader();
         InvocationManager invMgr = serverContext.getInvocationManager();
         boolean invocationCreated = false;
         if (invMgr.getCurrentInvocation() == null && capturedInvocation != null) {
@@ -98,10 +97,11 @@ public class JavaEEContextUtilImpl implements JavaEEContextUtil, Serializable {
             invMgr.preInvoke(newInvocation);
             invocationCreated = true;
         }
+        ClassLoader oldClassLoader = null;
         if (invocationCreated) {
-            Utility.setContextClassLoader(getInvocationClassLoader());
+            oldClassLoader = Utility.setContextClassLoader(getInvocationClassLoader());
         }
-        return new ContextImpl.Context(oldClassLoader, invocationCreated? invMgr.getCurrentInvocation() : null, invMgr);
+        return new ContextImpl.Context(invocationCreated? invMgr.getCurrentInvocation() : null, invMgr, oldClassLoader);
     }
 
     /**
@@ -114,7 +114,7 @@ public class JavaEEContextUtilImpl implements JavaEEContextUtil, Serializable {
     public Context pushRequestContext() {
         Context rootCtx = pushContext();
         BoundRequestContext brc = CDI.current().select(BoundRequestContext.class).get();
-        ContextImpl.RequestContext context = new ContextImpl.RequestContext(rootCtx, brc.isActive()? null : brc, new HashMap<String, Object>());
+        ContextImpl.RequestContext context = new ContextImpl.RequestContext(rootCtx, brc.isActive()? null : brc, new HashMap<>());
         if (context.ctx != null) {
             context.ctx.associate(context.storage);
             context.ctx.activate();
@@ -194,10 +194,16 @@ public class JavaEEContextUtilImpl implements JavaEEContextUtil, Serializable {
     }
 
     private void createInvocationContext() {
-        capturedInvocation = new ComponentInvocation();
-        capturedInvocation.componentId = instanceComponentId;
-        capturedInvocation.setJNDIEnvironment(compEnvMgr.getJndiNameEnvironment(instanceComponentId));
-        capturedInvocation.setComponentInvocationType(ComponentInvocation.ComponentInvocationType.SERVLET_INVOCATION);
+        JndiNameEnvironment jndiEnv = compEnvMgr.getJndiNameEnvironment(instanceComponentId);
+        if (jndiEnv != null) { // create invocation only for valid JNDI environment
+            capturedInvocation = new ComponentInvocation();
+            capturedInvocation.componentId = instanceComponentId;
+            capturedInvocation.setJNDIEnvironment(jndiEnv);
+            capturedInvocation.setComponentInvocationType(ComponentInvocation.ComponentInvocationType.SERVLET_INVOCATION);
+        }
+        else {
+            capturedInvocation = null;
+        }
     }
 
 

--- a/appserver/web/gf-web-connector/src/main/java/fish/payara/appserver/context/JavaEEContextUtilImpl.java
+++ b/appserver/web/gf-web-connector/src/main/java/fish/payara/appserver/context/JavaEEContextUtilImpl.java
@@ -69,6 +69,14 @@ import org.jvnet.hk2.annotations.Service;
 @Service
 @PerLookup
 public class JavaEEContextUtilImpl implements JavaEEContextUtil, Serializable {
+    private transient @Getter(AccessLevel.PACKAGE) ServerContext serverContext;
+    private transient ComponentEnvManager compEnvMgr;
+    private transient ComponentInvocation capturedInvocation;
+    private @Getter(onMethod = @__(@Override)) String instanceComponentId;
+    private static final String EMPTY_COMPONENT = "___EMPTY___";
+    private static final long serialVersionUID = 1L;
+
+
     @PostConstruct
     void init() {
         serverContext = Globals.getDefaultHabitat().getService(ServerContext.class);
@@ -99,7 +107,12 @@ public class JavaEEContextUtilImpl implements JavaEEContextUtil, Serializable {
         }
         ClassLoader oldClassLoader = null;
         if (invocationCreated) {
-            oldClassLoader = Utility.setContextClassLoader(getInvocationClassLoader());
+            if (EMPTY_COMPONENT.equals(getInstanceComponentId())) {
+                oldClassLoader = Utility.getClassLoader();
+            }
+            else {
+                oldClassLoader = Utility.setContextClassLoader(getInvocationClassLoader());
+            }
         }
         return new ContextImpl.Context(invocationCreated? invMgr.getCurrentInvocation() : null, invMgr, oldClassLoader);
     }
@@ -153,6 +166,12 @@ public class JavaEEContextUtilImpl implements JavaEEContextUtil, Serializable {
     }
 
     @Override
+    public void setEmptyInvocation() {
+        instanceComponentId = EMPTY_COMPONENT;
+        capturedInvocation = createInvocation(null);
+    }
+
+    @Override
     public String getInvocationComponentId() {
         ComponentInvocation inv = serverContext.getInvocationManager().getCurrentInvocation();
         return inv != null? inv.getComponentId() : null;
@@ -196,20 +215,18 @@ public class JavaEEContextUtilImpl implements JavaEEContextUtil, Serializable {
     private void createInvocationContext() {
         JndiNameEnvironment jndiEnv = compEnvMgr.getJndiNameEnvironment(instanceComponentId);
         if (jndiEnv != null) { // create invocation only for valid JNDI environment
-            capturedInvocation = new ComponentInvocation();
-            capturedInvocation.componentId = instanceComponentId;
-            capturedInvocation.setJNDIEnvironment(jndiEnv);
-            capturedInvocation.setComponentInvocationType(ComponentInvocation.ComponentInvocationType.SERVLET_INVOCATION);
+            capturedInvocation = createInvocation(jndiEnv);
         }
         else {
             capturedInvocation = null;
         }
     }
 
-
-    private transient @Getter(AccessLevel.PACKAGE) ServerContext serverContext;
-    private transient ComponentEnvManager compEnvMgr;
-    private transient ComponentInvocation capturedInvocation;
-    private @Getter(onMethod = @__(@Override)) String instanceComponentId;
-    private static final long serialVersionUID = 1L;
+    private ComponentInvocation createInvocation(JndiNameEnvironment jndiEnv) {
+        ComponentInvocation newInvocation = new ComponentInvocation();
+        newInvocation.componentId = instanceComponentId;
+        newInvocation.setJNDIEnvironment(jndiEnv);
+        newInvocation.setComponentInvocationType(ComponentInvocation.ComponentInvocationType.SERVLET_INVOCATION);
+        return newInvocation;
+    }
 }

--- a/nucleus/common/internal-api/src/main/java/org/glassfish/internal/api/JavaEEContextUtil.java
+++ b/nucleus/common/internal-api/src/main/java/org/glassfish/internal/api/JavaEEContextUtil.java
@@ -100,6 +100,12 @@ public interface JavaEEContextUtil {
      */
     String getInstanceComponentId();
 
+    /**
+     * Set a valid component invocation that's empty,
+     * i.e. doesn't belong to any module
+     */
+    void setEmptyInvocation();
+
     interface Context extends Closeable {};
     interface Closeable extends AutoCloseable {
         @Override

